### PR TITLE
feat(tuner): enhance tuner with prompt tuning and model selection

### PR DESCRIPTION
**Pull request by Kanaricc**
## AgentScope Version

1.0.13

## Description

Enhance tuner with prompt tuning and model selection.

- Integrate DSPy to optimize agent system prompts in the workflow.
- Introduce model selection based on performance, latency, and cost.
- Reuse existing workflow, judge, and model interfaces to minimize changes.

## Checklist

Please check the following items before code is ready to be reviewed.

- [x]  Code has been formatted with `pre-commit run --all-files` command
- [x]  All tests are passing
- [x]  Docstrings are in Google style
- [x]  Related documentation has been updated (e.g. links, examples, etc.)
- [x]  Code is ready for review

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -40,6 +40,10 @@ dependencies = [
     "sounddevice",
     "sqlalchemy",
     "python-frontmatter",
+    "trinity-rft>=0.2.1",
```
Comment (pan-x-c) [inline pyproject.toml]: 1. move them to optional dependencies
2. trinity-rft>=0.2.1 -> trinity-rft>=0.4.1

File: src/agentscope/tuner/model_selection/_model_selection.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    Returns:
+        `ChatModelBase`: The model that achieved the best performance across
+                         the dataset.
+                         For custom judge functions (higher reward is better),
+                         the model with the highest average reward is returned.
+                         For built-in judge functions (lower consumption is
+                         better), the model with the lowest average (most
+                         negative) reward is returned.
+    """
+    if not candidate_models:
```
Comment (pan-x-c) [inline src/agentscope/tuner/model_selection/_model_selection.py]: -> len(candidate_models) > 1

File: src/agentscope/tuner/model_selection/_model_selection.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+    logger.info(
+        "Evaluating %d candidate models: %s",
+        len(candidate_models),
+        [model.model_name for model in candidate_models],
+    )
+
+    # Determine if we're using built-in judge functions that favor lower values
+    # We check if the judge function is one of the known built-in functions
+    is_minimization_task = (
```
Comment (pan-x-c) [inline src/agentscope/tuner/model_selection/_model_selection.py]: Using names to determine task types is prone to errors. We should use a more robust method (e.g., decorator).

File: src/agentscope/tuner/model_selection/_model_selection.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+            None.
+
+    Returns:
+        `ChatModelBase`: The model that achieved the best performance across
+                         the dataset.
+                         For custom judge functions (higher reward is better),
+                         the model with the highest average reward is returned.
+                         For built-in judge functions (lower consumption is
+                         better), the model with the lowest average (most
+                         negative) reward is returned.
```
Comment (pan-x-c) [inline src/agentscope/tuner/model_selection/_model_selection.py]: It's a bit confusing that there are two different optimization targets.
The optimization direction should be unified, and only one of them should be exposed to users.

File: src/agentscope/tuner/model_selection/_model_selection.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+            workflow_output.metrics["execution_time"] = execution_time
+
+            # Evaluate the workflow output using judge function
+            judge_output: JudgeOutput = await judge_func(
+                sample,
+                workflow_output,
+                auxiliary_models or {},
+            )
+
+            total_reward += judge_output.reward
```
Comment (pan-x-c) [inline src/agentscope/tuner/model_selection/_model_selection.py]: It would be btter to record and return metrics generated by judge.

File: src/agentscope/tuner/model_selection/_model_selection.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+
+
+async def select_model(
+    *,
+    workflow_func: WorkflowType,
+    judge_func: JudgeType,
+    train_dataset: DatasetConfig,
+    candidate_models: Sequence[ChatModelBase],
+    auxiliary_models: dict[str, ChatModelBase] | None = None,
+) -> ChatModelBase:
```
Comment (pan-x-c) [inline src/agentscope/tuner/model_selection/_model_selection.py]: -> Tuple[ChatModelBase, Dict[str, float]]

Return some metrics to provide a detailed description of the reasons for the selection.

File: examples/tuner/prompt_tuning/example.py [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+    Returns:
+        WorkflowType: A workflow function that takes task, model,
+            and auxiliary_models as arguments and returns a WorkflowOutput.
+    """
+
+    async def run_react_agent(
+        task: dict,
+        model: ChatModelBase,
+        auxiliary_models: dict[str, ChatModelBase],
```
Comment (pan-x-c) [inline examples/tuner/prompt_tuning/example.py]: In prompt tuning, `model` and `auxiliary_models` are not necessary and can be removed, and `prompt` should be a required field.

File: examples/tuner/prompt_tuning/example.py [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+            response=response,
+        )
+
+    return run_react_agent
+
+
+async def gsm8k_judge(
+    task: dict,
+    response: Msg,
+    auxiliary_models: dict[str, ChatModelBase],
```
Comment (pan-x-c) [inline examples/tuner/prompt_tuning/example.py]: `auxiliary_models` can be removed, its only for model tuning

File: examples/tuner/prompt_tuning/README.md [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+
+        return WorkflowOutput(
+            response=response,
+        )
+
+    return run_react_agent
+
+
+async def judge_function(
+    task: Dict, response: Msg, auxiliary_models: Dict[str, ChatModelBase]
```
Comment (pan-x-c) [inline examples/tuner/prompt_tuning/README.md]: A similar issue, `auxiliary_models` can be removed

File: pyproject.toml [hunk 10]
```diff
@@ -137,6 +137,14 @@ rag = [
 # ------------ Evaluation ------------
 evaluate = ["ray"]
 
+# ------------ Tuner ------------
+tuner = [
+    "trinity-rft>=0.4.1",
+    "dspy>=3.1.0",
+    "datasets>=2.20.0",
```
Comment (pan-x-c) [inline pyproject.toml]:
```suggestion
    "datasets>=4.0.0",
```

File: pyproject.toml [hunk 11]
```diff
@@ -137,6 +137,14 @@ rag = [
 # ------------ Evaluation ------------
 evaluate = ["ray"]
 
+# ------------ Tuner ------------
+tuner = [
+    "trinity-rft>=0.4.1",
```
Comment (pan-x-c) [inline pyproject.toml]:
```suggestion
    "trinity-rft>=0.5.0",
```

File: src/agentscope/tuner/_workflow.py [hunk 12]
```diff
@@ -54,3 +54,21 @@ class WorkflowOutput(BaseModel):
 #     `WorkflowOutput`:
 #         The workflow execution results, including optional reward, raw
 #         response and metrics.
+
+
+PromptTuneWorkflowType = Callable[
+    [Dict, str],
+    Awaitable[WorkflowOutput],
+]
```
Comment (pan-x-c) [inline src/agentscope/tuner/_workflow.py]:
```suggestion
WorkflowType = Callable[..., Awaitable[WorkflowOutput]]
```

File: examples/tuner/prompt_tuning/example.py [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+async def gsm8k_judge(
+    task: dict,
+    response: Msg,
+) -> JudgeOutput:
+    """A simple judge function to calculate reward based on agent's response.
+
+    Args:
+        task (Dict): The task information for the corresponding workflow.
+        response (Msg): The response generated by the corresponding workflow.
+        auxiliary_models (Dict[str, ChatModelBase]):
```
Comment (pan-x-c) [inline examples/tuner/prompt_tuning/example.py]: Remove stale comments

File: src/agentscope/tuner/_workflow.py [hunk 14]
```diff
@@ -42,6 +37,8 @@ class WorkflowOutput(BaseModel):
 #     model (`ChatModelBase`):
 #         The primary chat model used in the workflow, this is the main model
 #         being tuned.
+#     system_prompt (`str`):
```
Comment (pan-x-c) [inline src/agentscope/tuner/_workflow.py:40]: Specific when this parameter is required, the same for other parameters

File: tests/model_selection_test.py
Comment (pan-x-c) [inline tests/model_selection_test.py:1]: The empty file can be removed

File: examples/tuner/model_selection/example_bleu.py [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+    response = await agent.reply(msg=msg)
+
+    return WorkflowOutput(
+        response=response,
+    )
+
+
+async def bleu_judge(
+    task: Dict[str, Any],
+    response: WorkflowOutput,
```
Comment (pan-x-c) [inline examples/tuner/model_selection/example_bleu.py]: Please align with model tuning. Only pass the `response: Any` field of `WorkflowOutput` here.

File: examples/tuner/model_selection/example_bleu.py [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+) -> JudgeOutput:
+    """A judge function to calculate BLEU score for translation quality.
+
+    Args:
+        task (Dict[str, Any]): The task information containing reference translation.
+        response: The response generated by the translation workflow (should be WorkflowOutput).
```
Comment (pan-x-c) [inline examples/tuner/model_selection/example_bleu.py]: Same as above, the comments need to be updated. Additionally, other built-in judges in `src/agentscope/tuner/model_selection/_built_in_judges.py` appear to have the same issue.

File: examples/tuner/prompt_tuning/README.md [hunk 17]
```diff
@@ -0,0 +1,348 @@
...
+    task: Dict,
+    response: Any,
+    auxiliary_models: Dict[str, ChatModelBase],
+) -> JudgeOutput:
+    """Calculate reward based on the input task and agent's response."""
+```
+
+- Inputs:
+    - `task`: A dictionary representing a single training task.
+    - `response`: The output from the workflow function.
```
Comment (pan-x-c) [inline examples/tuner/prompt_tuning/README.md]:
```suggestion
    - `response`: The `response` field of the `WorkflowOutput` struct returned by the workflow function.
```

File: examples/tuner/prompt_tuning/README.md [hunk 18]
```diff
@@ -0,0 +1,348 @@
...
+
+### Step 3: Implement the judge function
+
+The judge function evaluates the agent's response and returns a reward. It has the same signature as in RL-based tuning.
+
+```python
+async def judge_function(
+    task: Dict,
+    response: Any,
+    auxiliary_models: Dict[str, ChatModelBase],
```
Comment (pan-x-c) [inline examples/tuner/prompt_tuning/README.md]: Prompt tuning does not need auxiliary models, and this filed can be removed

File: pyproject.toml [hunk 19]
```diff
@@ -137,6 +137,14 @@ rag = [
 # ------------ Evaluation ------------
 evaluate = ["ray"]
 
+# ------------ Tuner ------------
+tuner = [
+    "trinity-rft>=0.5.0",
```
Comment (pan-x-c) [inline pyproject.toml:148]: `trinity-rft` is too heavy, which can be moved into a separate `tuner-gpu` dependency.

File: README_zh.md [hunk 20]
```diff
@@ -356,7 +356,7 @@ asyncio.run(multi_agent_conversation())
 
 ### 微调
 
-- [调优 ReAct 智能体](https://github.com/agentscope-ai/agentscope/tree/main/examples/tuner/react_agent)
+- [调优 ReAct 智能体](https://github.com/agentscope-ai/agentscope/tree/main/examples/tuner/model_tuning)
```
Comment (pan-x-c) [inline README_zh.md:359]: Please also update `docs/tutorial/en/src/task_tuner.py` and `docs/tutorial/zh_CN/src/task_tuner.py`

File: pyproject.toml [hunk 21]
```diff
@@ -139,6 +137,22 @@ rag = [
 # ------------ Evaluation ------------
 evaluate = ["ray"]
 
+# ------------ Tuner ------------
+tuner = [
+    "dspy>=3.1.0",
+    "datasets>=4.0.0",
+    "litellm[proxy]>=1.75.3",
+    datasets,
```
Comment (pan-x-c) [inline pyproject.toml]: remove `datasets` and `sacrebleu`

File: pre-commit_log.md
Comment (pan-x-c) [inline pre-commit_log.md:1]: remove this file
